spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java

```diff
@@ -0,0 +1,188 @@
+
+	private BackOffPolicy backOffPolicy = new FixedBackOffPolicy(1000);
+
+	private RetryListener retryListener = new RetryListener() {
+	};
+
+	/**
+	 * Create a new retry template with default settings.
+	 */
+	public RetryTemplate() {
+	}
+
+	/**
+	 * Create a new retry template with a custom {@link RetryPolicy}.
+	 * @param retryPolicy the retry policy to use
+	 */
+	public RetryTemplate(RetryPolicy retryPolicy) {
+		Assert.notNull(retryPolicy, "Retry policy must not be null");
+		this.retryPolicy = retryPolicy;
+	}
+
+	/**
+	 * Create a new retry template with a custom {@link RetryPolicy} and {@link BackOffPolicy}.
+	 * @param retryPolicy the retry policy to use
+	 * @param backOffPolicy the backoff policy to use
+	 */
+	public RetryTemplate(RetryPolicy retryPolicy, BackOffPolicy backOffPolicy) {
+		this(retryPolicy);
+		Assert.notNull(backOffPolicy, "BackOff policy must not be null");
+		this.backOffPolicy = backOffPolicy;
+	}
+
+	/**
+	 * Set the {@link RetryPolicy} to use. Defaults to <code>MaxAttemptsRetryPolicy()</code>.
+	 * @param retryPolicy the retry policy to use. Must not be <code>null</code>.
+	 */
+	public void setRetryPolicy(RetryPolicy retryPolicy) {
+		Assert.notNull(retryPolicy, "Retry policy must not be null");
+		this.retryPolicy = retryPolicy;
+	}
+
+	/**
+	 * Set the {@link BackOffPolicy} to use. Defaults to <code>FixedBackOffPolicy(Duration.ofSeconds(1))</code>.
+	 * @param backOffPolicy the backoff policy to use. Must not be <code>null</code>.
+	 */
+	public void setBackOffPolicy(BackOffPolicy backOffPolicy) {
+		Assert.notNull(backOffPolicy, "BackOff policy must not be null");
+		this.backOffPolicy = backOffPolicy;
+	}
+
+	/**
+	 * Set the {@link RetryListener} to use. Defaults to a <code>NoOp</code> implementation.
+	 * If multiple listeners are needed, use a {@link CompositeRetryListener}.
+	 * @param retryListener the retry listener to use. Must not be <code>null</code>.
+	 */
+	public void setRetryListener(RetryListener retryListener) {
+		Assert.notNull(retryListener, "Retry listener must not be null");
+		this.retryListener = retryListener;
+	}
+
+	/**
+	 * Call the retry callback according to the configured retry and backoff policies.
+	 * If the callback succeeds, its result is returned. Otherwise, a {@link RetryException}
+	 * will be thrown to the caller having all attempt exceptions as suppressed exceptions.
+	 * @param retryCallback the callback to call initially and retry if needed
+	 * @param <R> the type of the result
+	 * @return the result of the callback if any
+	 * @throws RetryException thrown if the retry policy is exhausted. All attempt exceptions
+	 * are added as suppressed exceptions to the final exception.
+	 */
+	@Override
+	public <R> @Nullable R execute(RetryCallback<R> retryCallback) throws RetryException {
+		Assert.notNull(retryCallback, "Retry Callback must not be null");
+		String callbackName = retryCallback.getName();
+		// initial attempt
+		try {
+			if (logger.isDebugEnabled()) {
```

Comment:
Not important.
Can we use a `LogAccessor` API in Spring Core itself to make this code more clean?

Reply:
I think it is not (only) about formatting, but rather about replacing all code blocks (`if (logger.isDebugEnabled()) { logger.debug(...); }`) with a single statement. I believe this is better for readability. Even if `LogAccessor` is not currently used in Spring Framework itself, may be this is an opportunity to start using it? I personally like that API and think to start using it in Batch.

For now, I updated the PR to use `LogAccessor`, but if the team thinks it's not appropriate, then I will revert that change.

Reply:
> I think it is not (only) about formatting, but rather about replacing all code blocks (`if (logger.isDebugEnabled()) { logger.debug(...); }`) with a single statement.

You don't have to use `LogAccessor` to achieve that.

We achieve that in multiple places using only `LogMessage`.

For example:

https://github.com/spring-projects/spring-framework/blob/b699b65b40a0299d4845c2d0160ce1a0145d85d4/spring-beans/src/main/java/org/springframework/beans/factory/support/AbstractBeanFactory.java#L1746-L1748